src/DI/ServiceLookup/CallSiteFactory.cs

```diff
@@ -206,6 +208,114 @@ private IServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type
+
+                if ((specialConstraints & GenericParameterAttributes.DefaultConstructorConstraint)
+                    != GenericParameterAttributes.None)
+                {
+                    if (!parameterTypeInfo.IsValueType && parameterTypeInfo.DeclaredConstructors.All(c => c.GetParameters().Length != 0))
+                    {
+                        return false;
+                    }
+                }
+
+                if ((specialConstraints & GenericParameterAttributes.ReferenceTypeConstraint)
+                    != GenericParameterAttributes.None)
+                {
+                    if (parameterTypeInfo.IsValueType)
+                    {
+                        return false;
+                    }
+                }
+
+                if ((specialConstraints & GenericParameterAttributes.NotNullableValueTypeConstraint)
+                    != GenericParameterAttributes.None)
+                {
+                    if (!parameterTypeInfo.IsValueType ||
+                        (parameterTypeInfo.IsGenericType && IsGenericTypeDefinedBy(parameter, typeof(Nullable<>))))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static bool IsGenericTypeDefinedBy(Type type, Type openGeneric)
+        {
+            return !type.GetTypeInfo().ContainsGenericParameters
+                       && type.GetTypeInfo().IsGenericType
+                       && type.GetGenericTypeDefinition() == openGeneric;
+        }
+
+        private static Type SubstituteGenericParameterConstraint(Type[] parameters, Type constraint)
+        {
+            if (!constraint.IsGenericParameter) return constraint;
+            return parameters[constraint.GenericParameterPosition];
+        }
+
+        private static bool ParameterCompatibleWithTypeConstraint(Type parameter, Type constraint)
+        {
+            return constraint.IsAssignableFrom(parameter) ||
+                   new[] { parameter, parameter.BaseType }
+                       .Concat(parameter.GetTypeInfo().ImplementedInterfaces)
+                       .Where(p => p != null)
+                       .Any(p => ParameterEqualsConstraint(p, constraint));
+        }
+
+        private static bool ParameterEqualsConstraint(Type parameter, Type constraint)
+        {
+            var genericArguments = parameter.GenericTypeArguments;
+            if (genericArguments.Length > 0 && constraint.GetTypeInfo().IsGenericType)
+            {
+                var typeDefinition = constraint.GetGenericTypeDefinition();
+                if (typeDefinition.GetTypeInfo().GenericTypeParameters.Length == genericArguments.Length)
+                {
+                    Type genericType;
+                    try
+                    {
+                        genericType = typeDefinition.MakeGenericType(genericArguments);
+                    }
+                    catch (Exception)
```

Comment:
Should we catch only the expected exceptions? From the documentation, that's 

> InvalidOperationException | The current type does not represent a generic type definition. That is, IsGenericTypeDefinition returns false.

So we could/should be testing `typeDefinition.IsGenericTypeDefinition` first.

> ArgumentNullException | typeArguments is null.-or-Any element of typeArguments is null.
ArgumentException | The number of elements in typeArguments is not the same as the number of type parameters in the current generic type definition.-or-Any element of typeArguments does not satisfy the constraints specified for the corresponding type parameter of the current generic type.-or-typeArguments contains an element that is a pointer type (Type.IsPointer returns true), a by-ref type (Type.IsByRef returns true), or Void.
NotSupportedException | The invoked method is not supported in the base class. Derived classes must provide an implementation.